File: scripts/deeplx/deeplx_client.py
import json
import time
import urllib.request
import urllib.error
from typing import Optional, Dict, Any

DEFAULT_DEEPLX_URL = "http://127.0.0.1:1188"

class DeepLXClient:
    """
    Cliente para a API de tradução DeepLX / DLX (OwO-Network/DLX).
    Suporta verificação de status, tradução de texto simples e retentativas automáticas.
    """
    def __init__(self, base_url: str = DEFAULT_DEEPLX_URL, timeout: int = 15, max_retries: int = 5, retry_delay: float = 2.0):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_delay = retry_delay

    def check_health(self) -> Dict[str, Any]:
        """
        Verifica a conexão com a API DeepLX no endpoint raiz (GET /).
        Retorna o JSON retornado pela API (ex: code 200, DLX Translation API message).
        """
        req = urllib.request.Request(
            f"{self.base_url}/",
            headers={"User-Agent": "DeepLX-Python-Client/1.0"}
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                data = resp.read().decode("utf-8")
                return json.loads(data)
        except Exception as e:
            return {"code": 500, "error": str(e)}

    def translate(self, text: str, source_lang: str = "auto", target_lang: str = "PT") -> str:
        """
        Traduz um texto utilizando POST /translate no servidor DeepLX.
        Aplica retentativa em caso de erro 429 (Too Many Requests) ou falhas temporárias de rede.
        """
        if not text or not text.strip():
            return text

        url = f"{self.base_url}/translate"
        payload = {
            "text": text,
            "source_lang": source_lang.upper(),
            "target_lang": target_lang.upper()
        }
        json_payload = json.dumps(payload).encode("utf-8")
        headers = {
            "Content-Type": "application/json",
            "User-Agent": "DeepLX-Python-Client/1.0"
        }

        delay = self.retry_delay
        for attempt in range(1, self.max_retries + 1):
            try:
                req = urllib.request.Request(url, data=json_payload, headers=headers)
                with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                    res_body = resp.read().decode("utf-8")
                    data = json.loads(res_body)

                    if isinstance(data, dict):
                        if data.get("code") == 200 and "data" in data:
                            return data["data"]
                        elif "data" in data and isinstance(data["data"], str):
                            return data["data"]
                        elif "translated_text" in data:
                            return data["translated_text"]

                    raise ValueError(f"Resposta inesperada do DeepLX: {data}")

            except urllib.error.HTTPError as err:
                if err.code == 429 and attempt < self.max_retries:
                    print(f"[DeepLX] Erro 429 (Limitação de taxa). Aguardando {delay:.1f}s (Tentativa {attempt}/{self.max_retries})...")
                    time.sleep(delay)
                    delay *= 2.0  # Backoff exponencial
                elif attempt < self.max_retries:
                    print(f"[DeepLX] HTTP Error {err.code}. Aguardando {delay:.1f}s...")
                    time.sleep(delay)
                else:
                    print(f"[DeepLX] Falha ao traduzir após {self.max_retries} tentativas: HTTP {err.code}")
                    return text
            except Exception as ex:
                if attempt < self.max_retries:
                    time.sleep(delay)
                else:
                    print(f"[DeepLX] Erro de conexão/processamento: {ex}")
                    return text

        return text

if __name__ == "__main__":
    client = DeepLXClient()
    print("--- Testando Conexão com DeepLX API ---")
    status = client.check_health()
    print("Status do Servidor:", status)
    
    if status.get("code") == 200:
        sample_text = "Hello world! DeepLX translation service is ready."
        print(f"\nTexto Original: {sample_text}")
        result = client.translate(sample_text, target_lang="PT")
        print(f"Tradução PT-BR: {result}")

File: scripts/deeplx/deeplx_client.js
/**
 * Cliente de tradução DeepLX / DLX em Node.js (OwO-Network/DLX)
 * Suporta retentativa com backoff exponencial para contornar erros 429.
 */

const DEFAULT_URL = 'http://127.0.0.1:1188';

class DeepLXClient {
  constructor(baseUrl = DEFAULT_URL, options = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.maxRetries = options.maxRetries || 5;
    this.retryDelay = options.retryDelay || 2000;
  }

  /**
   * Verifica a saúde da API DeepLX no endpoint raiz (GET /)
   */
  async checkHealth() {
    try {
      const response = await fetch(`${this.baseUrl}/`);
      const data = await response.json();
      return data;
    } catch (err) {
      return { code: 500, error: err.message };
    }
  }

  /**
   * Traduz um texto para o idioma desejado (ex: PT / PT-BR)
   */
  async translate(text, sourceLang = 'auto', targetLang = 'PT') {
    if (!text || !text.trim()) return text;

    const url = `${this.baseUrl}/translate`;
    const body = JSON.stringify({
      text: text,
      source_lang: sourceLang.toUpperCase(),
      target_lang: targetLang.toUpperCase(),
    });

    let currentDelay = this.retryDelay;

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        const res = await fetch(url, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: body,
        });

        if (res.status === 429) {
          console.warn(`[DeepLX] 429 Rate Limit (tentativa ${attempt}/${this.maxRetries}). Aguardando ${currentDelay}ms...`);
          await new Promise((resolve) => setTimeout(resolve, currentDelay));
          currentDelay *= 2;
          continue;
        }

        if (!res.ok) {
          throw new Error(`HTTP Error ${res.status}: ${res.statusText}`);
        }

        const data = await res.json();
        if (data.code === 200 && data.data) {
          return data.data;
        } else if (data.data && typeof data.data === 'string') {
          return data.data;
        }

        throw new Error(`Resposta inesperada: ${JSON.stringify(data)}`);
      } catch (err) {
        if (attempt === this.maxRetries) {
          console.error(`[DeepLX] Falha ao traduzir após ${this.maxRetries} tentativas:`, err.message);
          return text;
        }
        await new Promise((resolve) => setTimeout(resolve, currentDelay));
      }
    }

    return text;
  }
}

module.exports = { DeepLXClient };

// Exemplo de uso se executado diretamente
if (require.main === module) {
  (async () => {
    const client = new DeepLXClient();
    console.log('--- Testando DeepLX Client JS ---');
    const health = await client.checkHealth();
    console.log('Status da API:', health);

    if (health.code === 200) {
      const translated = await client.translate('Hello world from Node.js DeepLX Client!', 'auto', 'PT');
      console.log('Tradução PT-BR:', translated);
    }
  })();
}
